### PR TITLE
Check DSA parameters for excessive sizes before validating

### PR DESCRIPTION
**Pull request by t8m**
This avoids overly long computation of various validation checks.

Fixes CVE-2024-4603
(Low severity)

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale please reconfirm, I was too eager with always checking the g value - it cannot be mandatory in the ACVP tests.

**Review by paulidale — APPROVED** [review]

File: crypto/dsa/dsa_check.c [hunk 1]
```diff
@@ -19,8 +19,34 @@
 #include "dsa_local.h"
 #include "crypto/dsa.h"
 
+static int dsa_precheck_params(const DSA *dsa, int *ret)
+{
+    if (dsa->params.p == NULL || dsa->params.q == NULL) {
```
Comment (slontis) [inline crypto/dsa/dsa_check.c:24]: I dont think this check for q being NULL is correct.. shouldnt it just be part of the check on line 36..

Comment (slontis) [reply crypto/dsa/dsa_check.c:24]: See ossl_dsa_check_priv_key(),
and the pairwise doesnt use q either.

Comment (t8m) [reply crypto/dsa/dsa_check.c:24]: But the q param is mandatory in DSA parameters (not DH parameters!). I do not see anything wrong with the precheck enforcing mandatory params.

Comment (t8m) [reply crypto/dsa/dsa_check.c:24]: And IMO whether it is placed here or at line 36 is a matter of taste. I prefer it like this.

Comment (slontis) [reply crypto/dsa/dsa_check.c:24]: I thought the point of the check was for bounds, not to enforce it being non NULL..

The difference is if (q != NULL && q < )

Comment (slontis) [reply crypto/dsa/dsa_check.c:24]: You should also get static analysis hits with your change, because of the check in  ossl_dsa_check_priv_key

Comment (t8m) [reply crypto/dsa/dsa_check.c:24]: I can remove that redundant check.

**Comment by t8m** [comment]
@paulidale please reconfirm again

File: CHANGES.md [hunk 2]
```diff
@@ -69,7 +69,26 @@ OpenSSL 3.4
...
+   be very slow.
+
+   Applications that use the functions EVP_PKEY_param_check() or
+   EVP_PKEY_public_check() to check a DSA public key or DSA parameters may
+   experience long delays. Where the key or parameters that are being checked
+   have been obtained from an untrusted source this may lead to a Denial of
+   Service.
+
+   To resolve this issue DSA keys larger than OPENSSL_DSA_MAX_MODULUS_BITS
+   will now fail the check immediately with an DSA_R_MODULUS_TOO_LARGE error
```
Comment (mattcaswell) [inline CHANGES.md:84]: s/an/a/

**Review by mattcaswell — APPROVED** [review]
Approved subject to the minor nit fix

File: crypto/dsa/dsa_check.c [hunk 3]
```diff
@@ -19,8 +19,35 @@
...
+        return 0;
+    }
+
+    if (BN_num_bits(dsa->params.p) > OPENSSL_DSA_MAX_MODULUS_BITS) {
+        ERR_raise(ERR_LIB_DSA, DSA_R_MODULUS_TOO_LARGE);
+        *ret = FFC_CHECK_INVALID_PQ;
+        return 0;
+    }
+
+    if (BN_num_bits(dsa->params.q) >= BN_num_bits(dsa->params.p)) {
```
[the remark this answers is not in the document]
Comment (t8m) [reply crypto/dsa/dsa_check.c:36]: q must never be larger or equal to p.

File: CHANGES.md (hunk 2 above)
Comment (t8m) [reply CHANGES.md]: fixup pushed

**Review by nhorman — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Review by slontis — APPROVED** [review]

**Comment by t8m** [comment]
Merged to all the active branches. Thank you for the reviews.

**Review by 5twelve — APPROVED** [review]

**Pull request closed without merging**